src/sentry/backup/validate.py

```diff
@@ -21,60 +23,97 @@ def validate(
+            """Assigns the next available ordinal to the supplied `obj` model."""
+
+            pk = obj["pk"]
+            model = obj["model"]
+            findings = []
+            if pk > self.max_seen_pk:
+                self.max_seen_pk = pk
+            else:
+                findings.append(
+                    ComparatorFinding(
+                        kind="UnorderedInput",
+                        on=InstanceID(model, self.next_ordinal),
+                        left_pk=pk if side == Side.left else None,
+                        right_pk=pk if side == Side.right else None,
+                        reason=f"""instances not listed in ascending `pk` order; `pk` {pk} is less than or equal to {self.max_seen_pk} which precedes it""",
+                    )
+                )
+
+            obj["ordinal"] = self.next_ordinal
+            self.next_ordinal += 1
+            return (obj["ordinal"], findings if findings else [])
+
+    OrdinalCounters = Dict[str, OrdinalCounter]
+    ModelMap = Dict[InstanceID, JSONData]
+
+    def build_model_map(
+        models: JSONData, side: Side, findings: ComparatorFindings
+    ) -> Tuple[ModelMap, OrdinalCounters]:
+        """Does two things in tandem: builds a map of InstanceID -> JSON model, and simultaneously builds a map of model name -> number of ordinals assigned."""
+
+        model_map: ModelMap = {}
+        ordinal_counters: OrdinalCounters = {}
+        for model in models:
+            model_name = model["model"]
+            ordinal_counters.setdefault(model_name, OrdinalCounter())
+            counter = ordinal_counters[model_name]
+            ordinal, found = counter.assign(model, side)
+            findings.extend(found)
+            id = InstanceID(model_name, ordinal)
+            model_map[id] = model
+        return (model_map, ordinal_counters)
+
     def json_lines(obj: JSONData) -> list[str]:
         """Take a JSONData object and pretty-print it as JSON."""
 
         return JSON_PRETTY_PRINTER.encode(obj).splitlines()
 
-    findings = ComparatorFindings([])
-    exp_models = {}
-    act_models = {}
-
     # Because we may be scrubbing data from the objects as we compare them, we may (optionally) make
     # deep copies to start to avoid potentially mangling the input data.
-    expect = deepcopy(expect)
-    actual = deepcopy(actual)
-
-    # Ensure that the actual JSON contains no duplicates.
-    exp_dupes = set()
-    for model in expect:
-        id = InstanceID(model["model"], model["pk"])
-        if id in exp_models and id not in exp_dupes:
-            findings.append(
-                ComparatorFinding("DuplicateEntry", id, "Duplicate entry in expected output")
-            )
-            exp_dupes.add(id)
-        else:
-            exp_models[id] = model
-    act_dupes = set()
-    for model in actual:
-        id = InstanceID(model["model"], model["pk"])
-        if id in act_models and id not in act_dupes:
+    left_data = deepcopy(expect)
```

Comment:
nit: I thought expect vs actual was a bit more clear than left vs right

Reply:
I could be convinced, but ultimately I think consistency is more important than the specific names.

I do think that "expect v actual" sounds more like a testing framework, rather than a diffing framework, which is what this is. Ex, git-merge uses the left v right language.